### PR TITLE
Next.js Localization & i18next Config

### PR DESCRIPTION
**Pull request by goplayoutside3**
## Packages:

app-project
lib-classifier
lib-react-components

## Purpose

This PR includes config for Next.js locale [routing](https://nextjs.org/docs/advanced-features/i18n-routing) and config for [i18next](https://www.i18next.com/) translation. It’s laying the groundwork for subsequent PR’s translating frontend components in app-project, lib-classifier, and lib-react-components. Similar config for app-content-pages will be in a follow-up PR.

ADR 40 details the decision to use Next.js routing for changing locale, and ADR 41 explains the decision to replace counterpart with i18next. More details about i18next packages can be found in PR #2592.

`LocaleSwitcher` is a component for choosing locale. For now, it is commented out because we don’t want to render a locale menu on production when translation features are not complete. `LocaleSwitcher` will eventually go through design approval, too.

## To Test

If you’d like to test LocaleSwitcher, comment it in inside `ProjectHeader` and run any project locally. However, the rest of this PR should have no effect on FEM functionality.

Translations project board: https://github.com/zooniverse/front-end-monorepo/projects/15

### CODE REVIEW

**Comment by ErikOstlund** [comment]
I think this is an amazing start and something we need to continue working on! Internationalization is crazy important to enable users across the globe to use Zooniverse.

## Thoughts

1. Translations
As of now, this is using a static dictionary which is good to test with and set the ground work. Moving forward, consider looking into implementing a translator framework like:

- [next-translate](https://github.com/vinissimus/next-translate)
- [Google Translate API](https://cloud.google.com/translate/docs)
- [LinguiJS](https://lingui.js.org/)


2. Routing
This looks good to me. I see no issues. I'd like to hear other opinions.

**Comment by goplayoutside3** [comment]
Some more thoughts:

Pros: 
- Storing the current locale as a subpath in the url could be a useful source of truth throughout the site, and make it easy to share a project link translated to a specific language. Also, configuration of supported locales could be shared between app-project and app-content-pages.

Cons: 
- A major consequence I recently considered is that not all packages in our monorepo are Next.js apps, and that limits the ability to utilize Next.js features throughout.
    - For example, lib-classifier and lib-react-components are not Next.js apps and therefore do not have access to next/router's `locale`. Locale would instead need to be passed down as a prop, or stored as part of the state tree.
    - If we implement Next.js localization in app-project, storing the current locale as a subpath in the url could be a source of truth throughout the site, but a library like `counterpart` will still be needed to access a dictionary of static strings in places where next/router cannot be imported.

**Comment by shaunanoordin** [comment]
First off, Delilah++ for spearheading this translations implementation. 👍

I don't have strong opinions for or against the current implementation, with my default opinion being: _"Let's try this out and see how we feel about it!"_

I _do_ however have a bunch of rambling thoughts. Feel free to read and/or ignore everything after this line!

--------

Regarding the scope of translations/the grand strategy for localisation...
- Are we sticking to the current _"language choices are project-specific"_ functionality? (EDIT: wait, you said _"the LocaleSwitcher ... will only render when a project has available translations in Panoptes API."_ so the answer is yep)
  - By this I mean, e.g., [Galaxy Zoo](https://www.zooniverse.org/projects/zookeeper/galaxy-zoo/?language=fr) has a language option (dropdown in the project header, similar to Delilah's LocalSwitcher in ProjectHeader) for 4 languages
  - The translations for GZ cover everything from the links (platform-wide) to the researcher's quotes (project-specific) 
- I ask because we _also_ have _platform-wide_ translations, for a variety of languages, but you don't really have a "language choice" (i.e. no dropdown) unless you know the hacks. e.g. https://www.zooniverse.org/about/faq?language=pt
- I know it's not in the scope of this experiment, but I've been thinking about what our grand strategy for localisation is. 

Regarding everything that's not `app-project`...
- If everyone else is happy with this PR, I say that the next thing that needs to be prototyped is getting the localisation data down to `lib-classifier`.
- Wait, is it easy to share local info with `app-content-pages`? I thought anything relying on an external CMS is going to be a special case for handling translations.

Regarding embedding the locale info in the subpath...
- So if I'm on English (default), my URL is `zooniverse.org/projects/darkeshard/test-project-2021` but when I'm on French, my URL is `zooniverse.org/projects/FRENCH/darkeshard/test-project-2021` . Cool, that looks good and I'm pleasantly surprised to learn that changing the routing is pretty easy in Next.js.
- I was going to ask if whether the locale info should be closer to the path root, `zooniverse.org/FRENCH/projects/darkeshard/test-project-2021`, but I'm not sure if that's going to bring up the discussion of platform-wide vs project-specific scope of translations. 🤨

**Comment by eatyourgreens** [comment]
We could pass a `locale` prop into the classifier here:
https://github.com/zooniverse/front-end-monorepo/blob/80de98c88966c0d6764d8d9ae4a4f0dcca7f4354/packages/lib-classifier/src/components/Classifier/Classifier.js#L69-L81

Side note: we should document the existing classifier props, maybe by adding inline docgen  comments.

Adding a locale individually to each component in the shared React Components library might be cumbersome, or hard to maintain, but a helper might work eg. `export default withLocale(MyComponent)`, where `withLocale` is responsible for adding localisation consistently to each component in the library.

**Comment by mcbouslog** [comment]
> This is great, thank you Delilah for doing this! The code implementation here looks nice and straightforward.
> 
> I think it would be great to use Next.js's internationalization.
> 
> Apologies if this has been covered or I think the answer is right in front of me, but currently projects that are utilizing the FEM still have their Talk, Collect, and Recents pages in PFE (https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess and https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess/talk). Would there be an issue jumping between those translated pages? I'm not sure if that could be solved with nginx or Next middleware or something else.

@eatyourgreens - what do you think about going from a translated FEM project home or Classify to a translated PFE Talk/Collect/Recents? I think there are technical solutions to convert the locale in url (FEM Next.js localization) to locale as query param (PFE localization), and vice versa, but am not sure.

**Comment by eatyourgreens** [comment]
That might be a good reason to carry on using URL query parameters. Try it out with the links here. 

https://www.zooniverse.org/projects/nora-dot-eisner/planet-hunters-tess?language=fr

By the way, check out #2531 for an example of rewriting NextJS page paths from URL query parameters.

**Comment by goplayoutside3** [comment]
All great points thanks ya'll!

I think the fact that Talk, Collect, and Recents are eventually going to be a part of FEM is something to keep in mind. If there's a relatively painless way to use Next's middleware in combo with nginx to convert FEM's locale subpath to/from PFE's language query param, our end goal could use Next's subpath localization for all aspects of the platform. My instinct is to use a `_middleware.js` file (a feature of Next 12) in app-project to accomplish that conversion. @eatyourgreens since you recently worked with FEM and nginx, do you agree? Vercel has some example middleware files such as this [one](https://github.com/vercel/examples/blob/main/edge-functions/i18n/pages/_middleware.ts).

In response to @shaunanoordin's point about platform-wide localization, I agree we'll need to make some UX decisions. Ideally, a user has a seamless experience across pages if they prefer to interact with the site in a language other than default English. In app-project, the ability to view a project in French is dependent on that project's available French translations written by volunteers. However, the ability to view French translations in app-content-pages is dependent on Zooniverse providing the translated strings. That difference makes me think we should handle locale switching differently between the two for now. For instance, we can build a `LocaleSwitcher` component that renders in app-project only if the project has available translations, but later development of FEM may include a switcher component inside `ZooHeader` for other parts of the platform.

@shaunanoordin Re: sharing locale info with app-content-pages - If a user navigates from app-project to app-content-pages, a preferred locale would not be saved automatically with this localization technique. The only i18n configuration that can be shared between the two Next.js apps are the properties inserted into `next.config.js` specifying an array of locale subpaths and our default locale. We would need to build preferred locale into a cookie or middleware.

**Comment by goplayoutside3** [comment]
Noting that the description of this PR has been updated accordingly after the frontend team's decisions on Next.js routing and i18next translation packages. ADR's 40 and 41 outline those decisions.

File: packages/lib-classifier/src/store/RootStore.js [hunk 1]
```diff
@@ -97,6 +98,10 @@ const RootStore = types
       onPatch(self, _onPatch)
     }
 
+    function setLocale (newLocale) {
```
Comment (mcbouslog) [inline packages/lib-classifier/src/store/RootStore.js:101]: Consider adding a related basic test

**Review by mcbouslog — APPROVED** [review]
LGTM! Great work, exciting to have this moving forward.

File: packages/lib-classifier/src/store/RootStore.js (hunk 1 above)
Comment (goplayoutside3) [reply packages/lib-classifier/src/store/RootStore.js:101]: Thanks for the note! I added tests for `setLocale()` to the RootStore

File: packages/lib-classifier/src/components/Classifier/Classifier.js [hunk 2]
```diff
@@ -6,6 +6,8 @@ import { persist } from 'mst-persist'
 import PropTypes from 'prop-types'
 import React, { useEffect, useMemo, useState } from 'react'
 import zooTheme from '@zooniverse/grommet-theme'
+import '../../translations/i18n'
```
[the remark this answers is not in the document]
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:9]: This import bundles the translation dictionary files via `i18n.js` and the syntax is specific to `react-i18next`. For Next.js apps, the equivalent in `next-i18next` is the `appWithTranslation()` HOC used in app-project.

[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:9]: Ah I see! Thanks

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:9]: MDN has an explanation of this syntax: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only

**Pull request merged**